reference/security: add certificate-based authentication doc (#1737)
pingcap/docs: dev/reference/security/cert-based-authentication.md

### Configure certificate on TiDB server

Modify the `security` section in the TiDB configuration file:

{{< copyable "" >}}

```
[security]
ssl-cert ="path/to/server-cert.pem"
ssl-key ="path/to/server-key.pem"
ssl-ca="path/to/ca-cert.pem"
```

Start TiDB logs. If the following information is displayed in the log, the configuration is successful:

```
[INFO] [server.go:264] ["secure connection is enabled"] ["client verification enabled"=true]
```

### Configure certificate on client

Configure the client so that the client uses the client key and certificate for login.

Taking the MySQL client as an example, you can use the newly created client certificate, client key and CA by specifying `ssl-cert`, `ssl-key`, and `ssl-ca`:

{{< copyable "shell-regular" >}}

```bash
mysql -utest -h0.0.0.0 -P4000 --ssl-cert /path/to/client-cert.new.pem --ssl-key /path/to/client-key.new.pem --ssl-ca /path/to/ca-cert.pem
```

## Configure the user certificate information for login verification

Connect TiDB using the client to configure the login authentication. You can configure the user certificate information for login verification in the following three methods:

> + Configure the certificate when creating a user (`create user`):

Reviewer: ```suggestion
+ Configure the certificate information to be verified when creating a user (`create user`):
```

    {{< copyable "sql" >}}

    ```sql
    create user 'u1'@'%'  require issuer '/C=US/ST=California/L=San Francisco/O=PingCAP Inc./OU=TiDB/CN=TiDB admin/emailAddress=s@pingcap.com' subject '/C=US/ST=California/L=San Francisco/O=PingCAP Inc./OU=TiDB/CN=tpch-user1/emailAddress=zz@pingcap.com' cipher 'TLS_AES_256_GCM_SHA384';
    grant all on *.* to 'u1'@'%';
    ```

> + Configure the certificate when granting privilege:

Reviewer: ```suggestion
+ Configure the certificate information to be verified when granting privileges:
```
s/privilege/privileges

    {{< copyable "sql" >}}

    ```sql
    create user 'u1'@'%';
    grant all on *.* to 'u1'@'%' require issuer '/C=US/ST=California/L=San Francisco/O=PingCAP Inc./OU=TiDB/CN=TiDB admin/emailAddress=s@pingcap.com' subject '/C=US/ST=California/L=San Francisco/O=PingCAP Inc./OU=TiDB/CN=tpch-user1/emailAddress=zz@pingcap.com' cipher 'TLS_AES_256_GCM_SHA384';
    ```

> + Configure the certificate when altering a user:

Reviewer: ```suggestion
+ Configure the certificate information to be verified when altering a user:
```

    {{< copyable "sql" >}}

    ```sql
    alter user 'u1'@'%' require issuer '/C=US/ST=California/L=San Francisco/O=PingCAP Inc./OU=TiDB/CN=TiDB admin/emailAddress=s@pingcap.com' subject '/C=US/ST=California/L=San Francisco/O=PingCAP Inc./OU=TiDB/CN=tpch-user1/emailAddress=zz@pingcap.com' cipher 'TLS_AES_256_GCM_SHA384';
    ```

> In the above three methods, `require subject`, `require issuer`, and `require cipher` are respectively used to check the X509 certificate attributes. You can configure one item or multiple items using the space or `and` as the separator.

Reviewer: ```suggestion
In the above three methods, `require subject`, `require issuer`, and `require cipher` are used to check the X509 certificate attributes. You can configure one item or multiple items using the space or `and` as the separator.
```

### Update CA key and certificate

1. Back up the old CA key and certificate (suppose that `ca-key.pem` is stolen):

    {{< copyable "shell-regular" >}}

    ```bash
    mv ca-key.pem ca-key.old.pem
    mv ca-cert.pem ca-cert.old.pem
    ```

2. Generate the new CA key:

    {{< copyable "shell-regular" >}}

    ```bash
    sudo openssl genrsa 2048 > ca-key.pem
    ```

3. Generate the new CA certificate using the newly generated CA key:

    {{< copyable "shell-regular" >}}

    ```bash
    sudo openssl req -new -x509 -nodes -days 365000 -key ca-key.pem -out ca-cert.new.pem
    ```

    > **Note:**
    >
    > Generating the new CA certificate is to replace the keys and certificates on the client and server, and to ensure that online users are not affected. Therefore, the appended information in the above command must be consistent with the `require issuer` information.

4. Generate the combined CA certificate:

    {{< copyable "shell-regular" >}}

    ```bash
    cat ca-cert.new.pem ca-cert.old.pem > ca-cert.pem
    ```

> After this update, use the combined CA certificate and restart the TiDB server. Then the server accepts both the new and old CA certificates.

Reviewer: ```suggestion
After the above operations, restart the TiDB server with the newly created combined CA certificate. Then the server accepts both the new and old CA certificates.
```

1. Generate the new RSA key of the client:

    {{< copyable "shell-regular" >}}

    ```bash
    sudo openssl req -newkey rsa:2048 -days 365000 -nodes -keyout client-key.new.pem -out client-req.new.pem
    sudo openssl rsa -in client-key.new.pem -out client-key.new.pem
    ```

    > **Note:**
    >
    > The above command is to replace the client key and certificate, and to ensure that the online users are not affected. Therefore, the appended information in the above command must be consistent with the `require subject` information.

2. Use the combined certificate and the new CA key to generate the new client certificate:

    {{< copyable "shell-regular" >}}

    ```bash
    sudo openssl x509 -req -in client-req.new.pem -days 365000 -CA ca-cert.pem -CAkey ca-key.pem -set_serial 01 -out client-cert.new.pem
    ```

3. Make the client (for example, MySQL) connect TiDB with the new client key and certificate:

    {{< copyable "shell-regular" >}}

    ```bash
    mysql -utest -h0.0.0.0 -P4000 --ssl-cert /path/to/client-cert.new.pem --ssl-key /path/to/client-key.new.pem --ssl-ca /path/to/ca-cert.pem
    ```

### Update the server key and certificate

> 1. Generate the new RSA key of the sever:

Reviewer: ```suggestion
1. Generate the new RSA key of the server:
```

    Signature ok
    subject=C = US, ST = California, L = San Francisco, O = PingCAP Inc., OU = TiDB, CN = tpch-user1, emailAddress = zz@pingcap.com
    Getting CA Private Key
    ```

    > **Note:**
    >
    > The information of the `subject` section in the above output is used for [certificate configuration for login verification](#configure-the-user-certificate-information-for-login-verification) in the `require` section.

### Verify certificate

Execute the following command to verify certificate:

{{< copyable "shell-regular" >}}

```bash
openssl verify -CAfile ca-cert.pem server-cert.pem client-cert.pem
```

If the certificate is verified, you will see the following result:

```
server-cert.pem: OK
client-cert.pem: OK
```

## Configure TiDB and the client to use certificates

After generating the certificates, you need to configure the TiDB server and the client to use the corresponding server certificate or client certificate.

### Configure TiDB to use server certificate

Modify the `security` section in the TiDB configuration file:

{{< copyable "" >}}

```
[security]
ssl-cert ="path/to/server-cert.pem"
ssl-key ="path/to/server-key.pem"
ssl-ca="path/to/ca-cert.pem"
```

> Start TiDB logs. If the following information is displayed in the log, the configuration is successful:

Reviewer 2: maybe better some thing like "start TiDB and watch logs"
